### PR TITLE
Remove `MultiOr` form `Result` and push branching into LogicT

### PR DESCRIPTION
**Pull request by goodlyrottenapple**

### CODE REVIEW

File: kore/test/Test/Kore/Rewrite/RewriteStep.hs [hunk 1]
```diff
@@ -954,14 +934,14 @@ test_applyRewriteRule_ =
         assertEqual "" expect actualAxiom
         assertEqual "" expect actualClaim
     , testCase "rule a => \\bottom" $ do
-        let expect = [OrPattern.fromPatterns []]
+        let expect = [pure (mkBottom Mock.testSort)]
```
Comment (goodlyrottenapple) [inline kore/test/Test/Kore/Rewrite/RewriteStep.hs:937]: `OrPattern.fromPatterns [] ~ #bottom`

File: kore/src/Kore/Rewrite/RewriteStep.hs [hunk 2]
```diff
@@ -423,6 +439,15 @@ applyRewriteRulesParallel
...
             results <- applyRulesParallel sideConditionWithDefinedSubterms rules initial
+            -- applyWithFinalizer
+            --     sideConditionWithDefinedSubterms
+            --     ( finalizeRulesParallel
+            --         sideConditionWithDefinedSubterms
+            --         RewriteRule
+            --         (finalizeAppliedRule sideConditionWithDefinedSubterms)
+            --     )
+            --     rules
+            --     initial
```
Comment (goodlyrottenapple) [inline kore/src/Kore/Rewrite/RewriteStep.hs:450]: remove this

File: kore/src/Kore/Rewrite.hs [hunk 3]
```diff
@@ -226,20 +227,24 @@ transitionRule rewriteGroups assumeInitialDefined = transitionRuleWorker
...
+    Result.Results
+        ( Conditional
+            RewritingVariableName
+            (RulePattern RewritingVariableName)
+        )
+        a ->
+    TransitionT (RewriteRule RewritingVariableName, Seq SimplifierTrace) Simplifier (ProgramState a)
 deriveResults Result.Results{results, remainders} =
-    if (null results || all (\Result.Result{result} -> null result) results) && null remainders
+    if null results && null remainders
```
Comment (goodlyrottenapple) [inline kore/src/Kore/Rewrite.hs:238]: we need to check if all results are #bottom

File: kore/src/Kore/Rewrite.hs [hunk 4]
```diff
@@ -226,20 +227,24 @@ transitionRule rewriteGroups assumeInitialDefined = transitionRuleWorker
...
   where
     addResults results' = asum (addResult <$> results')
     addResult Result.Result{appliedRule, result} = do
-        (_, rules :: Seq SimplifierTrace) <- lift get
+        (_, simplifyRules :: Seq SimplifierTrace) <- lift get
         lift $ modify $ \(cache, _rules) -> (cache, mempty)
-        addRule (RewriteRule $ extract appliedRule, rules)
-        asum (pure . Rewritten <$> toList result)
+        addRule (RewriteRule $ extract appliedRule, simplifyRules)
+        pure . Rewritten $ result
```
Comment (goodlyrottenapple) [inline kore/src/Kore/Rewrite.hs:247]: return `empty` if `result` is `#bottom`

File: kore/src/Kore/Reachability/Claim.hs [hunk 5]
```diff
@@ -1036,17 +1036,15 @@ deriveResults ::
         simplifier
         (ApplyResult (Pattern RewritingVariableName))
 -- TODO (thomas.tuegel): Remove claim argument.
-deriveResults sort fromAppliedRule Results{results, remainders} =
+deriveResults _sort fromAppliedRule Results{results, remainders} =
```
Comment (goodlyrottenapple) [inline kore/src/Kore/Reachability/Claim.hs:1039]: get rid of `_sort`?

File: kore/src/Kore/Reachability/Claim.hs [hunk 6]
```diff
@@ -1036,17 +1036,15 @@ deriveResults ::
         simplifier
         (ApplyResult (Pattern RewritingVariableName))
 -- TODO (thomas.tuegel): Remove claim argument.
```
Comment (goodlyrottenapple) [inline kore/src/Kore/Reachability/Claim.hs:1038]:
```suggestion
```

File: kore/src/Kore/Rewrite/Result.hs [hunk 7]
```diff
@@ -122,16 +118,16 @@ remainder config = mempty{remainders = MultiOr.singleton config}
...
     MultiOr config
-gatherResults = foldMap result . results
+gatherResults = from . map result . toList . results
 
 -- | Distribute the 'Result' over a transition rule.
 transitionResult :: Result rule config -> TransitionT rule m config
 transitionResult Result{appliedRule, result} = do
     Transition.addRule appliedRule
-    asum (return <$> toList result)
+    return result
```
Comment (goodlyrottenapple) [inline kore/src/Kore/Rewrite/Result.hs:130]: should we discard #bottom her as well?

**Comment by goodlyrottenapple** [comment]
Tested this branch on KEVM integration tests to ensure that there is no functional change

**Review by geo2a — APPROVED** [review]
The unit test run need to be re-triggered once other tests have passed, as it's hit a pattern generation error (long standing issue with these unit tests that we've agreed to not address).

**Pull request merged**